### PR TITLE
feat(Forms) : rendering lib based on json-schema-form-core

### PR DESCRIPTION
**Pull request by jmfrancois**
**What is the problem this PR is trying to solve?**
The actual form implementation is based on mozilla's react-jsonschema-form lib which is too restrictive and miss a lot of features.

The biggest point is that it ties jsonSchema (which should represent the model), with the uiSchema format (which should represent the form, what widget, where it should appear).

For example, if we want a tab widget to manage a fieldset by tab, the jsonSchema (so the model) will be something like that : 
```javascript
{
    tab1: {
        lastname: '',
        firstname: '',
    },
    tab2: {
        address: '',
        city: '',
    }
}
```

If we want flat data but displayed with tabs, that is not possible. This introduce issues with daikkon forms.

**What is the chosen solution to this problem?**

Reimplement a lib based json-schema-form-core for jsonSchema/uiSchema parsing.

You can start by reading the [README](https://github.com/Talend/ui/tree/jmfrancois/forms-uischema-refactor/packages/forms/src/UIForm)

**Please check if the PR fulfills these requirements**
- [x] The PR commit message follows our [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)
- [ ] Related design / discussions / pages (not in jira), if any, are all linked or available in the PR

**[ ] This PR introduces a breaking change**

### CODE REVIEW

**Comment by jmfrancois** [comment]
waiting for https://github.com/json-schema-form/json-schema-form-core/issues/4 to continue

**Comment by mmaillart** [comment]
Some css color appear to be incorrect. The UI specs states this

- Label (when no value is entered): font-style : 'text', color : $dove-gray
- Value: font-style : 'text', color: $black
- Floating label (when a value is entered): font-style : 'small text', color :  $silver
- Hint text: font-style :'search boxes', color: $silver

File: packages/forms/src/UIForm/UIForm.component.js [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+const TRIGGER_AFTER = 'after';
+
+export default class UIForm extends React.Component {
+	constructor(props) {
+		super(props);
+		const { jsonSchema, uiSchema } = props;
+		this.state = {
+			mergedSchema: merge(jsonSchema, uiSchema),
+		};
+		console.log(this.state.mergedSchema)
```
Comment (talend-sromain) [inline packages/forms/src/UIForm/UIForm.component.js]: remove console log

File: output/forms.eslint.txt [hunk 2]
```diff
@@ -114,7 +114,14 @@ The react/require-extension rule is deprecated. Please use the import/extensions
   71:21  error  Missing trailing comma                                    comma-dangle
   77:16  error  'formContext' PropType is defined but prop is never used  react/no-unused-prop-types
 
-✖ 100 problems (100 errors, 0 warnings)
+/home/travis/build/Talend/ui/packages/forms/src/UIForm/actions/model.actions.js
+  3:1  error  Prefer default export  import/prefer-default-export
```
Comment (talend-sromain) [inline output/forms.eslint.txt]: some lints

File: packages/forms/src/UIForm/UIForm.component.js [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+		event.preventDefault();
+		const { mergedSchema } = this.state;
+		const { formName, properties, customValidation } = this.props;
+		const errors = validateAll(mergedSchema, properties, customValidation);
+
+		const isValid = !Object.keys(errors).length;
+		if (isValid) {
+			this.props.onSubmit(event, properties);
+		} else {
+			this.props.onValidateAll(formName, errors);
```
Comment (talend-sromain) [inline packages/forms/src/UIForm/UIForm.component.js]: Should we validateAll in all cases on submit ?
Validation here is not done on blur but as it may be later it can not validate field when there is some errors & enter key is pressed to submit the form

File: packages/forms/src/UIForm/UIForm.component.js [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+		onTrigger: PropTypes.func,
+		/** Custom widgets */
+		widgets: PropTypes.object, // eslint-disable-line react/forbid-prop-types
+
+		/** State management impl: The change callback */
+		onChange: PropTypes.func.isRequired,
+		/** State management impl: The form change callback */
+		onFormChange: PropTypes.func.isRequired,
+		/** State management impl: Partial fields validation callback */
+		onValidate: PropTypes.func,
```
Comment (talend-sromain) [inline packages/forms/src/UIForm/UIForm.component.js]: Rename on validate to setError ?

File: packages/forms/src/UIForm/UIForm.component.js [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+		/** State management impl: All fields validation callback */
+		onValidateAll: PropTypes.func,
```
Comment (talend-sromain) [inline packages/forms/src/UIForm/UIForm.component.js]: Rename onValidateAll to setErrors ?

File: packages/forms/src/UIForm/UIForm.component.js [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+			this.props.onSubmit(event, properties);
+		} else {
+			this.props.onValidateAll(formName, errors);
+		}
+	}
+
+	render() {
+		const { autoComplete, errors, formName, properties, widgets } = this.props;
+		return (
+			<form onSubmit={this.submit} autoComplete={autoComplete}>
```
Comment (talend-sromain) [inline packages/forms/src/UIForm/UIForm.component.js]: add some props to let some customisation on form tag
should we add novalidate by default ?

File: packages/forms/src/UIForm/reducers/form.reducer.js [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+	TF_MUTATE_VALUE,
+	TF_VALIDATE_ALL,
+	TF_VALIDATE_PARTIAL,
+} from '../actions';
+import { omit } from '../utils/properties';
+import modelReducer from './model.reducer';
+import validationsReducer from './validations.reducer';
+
+/**
+ * Form reducer, that manage multiple form state.
```
Comment (talend-sromain) [inline packages/forms/src/UIForm/reducers/form.reducer.js]: Is this comment relevant ?

File: packages/forms/src/UIForm/UIForm.component.js (hunk 6 above)
Comment (jsomsanith) [inline packages/forms/src/UIForm/UIForm.component.js]: Added noValidate. Since we make static validations, this is covered

File: packages/forms/src/UIForm/actions/form.actions.test.js [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+const errors = { field: 'errors' };
+
+describe('Form actions', () => {
+	describe('#createForm action', () => {
+		it('should create the action payload', () => {
+			// when
+			const resultAction = createForm(formName, jsonSchema, uiSchema, properties, errors);
+
+			// then
+			expect(resultAction).toMatchSnapshot();
```
Comment (frassinier) [inline packages/forms/src/UIForm/actions/form.actions.test.js]: I'm not convinced by this kind of test here :/

File: packages/forms/src/UIForm/fields/Button.js [hunk 9]
```diff
@@ -0,0 +1,40 @@
+import React, { PropTypes } from 'react';
+
+import Message from '../Message';
+
+export default function Button(props) {
+	const { id, errorMessage, isValid, onTrigger, schema } = props;
+	const { description, title, triggers, type } = schema;
+
+	return (
+		<div className={isValid ? null : 'has-error'}>
```
Comment (frassinier) [inline packages/forms/src/UIForm/fields/Button.js]:
```
<div className={classNames({ 'has-error': !isValid })}>
```

File: packages/forms/src/UIForm/fields/Button.js [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+			<button
+				id={id}
+				className={'btn'}
```
Comment (frassinier) [inline packages/forms/src/UIForm/fields/Button.js]: This should be a parameter I guess

File: packages/forms/src/UIForm/fieldsets/Tabs.js [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+import theme from './Tabs.scss';
+
+export default function Tabs(props) {
+	const { schema, ...restProps } = props;
+	const tabs = schema.items;
+
+	return (
+		<RBTabs className={theme['tf-tabs']}>
+			{tabs.map((tabSchema, index) => {
+				const tabClassName = isValid(tabSchema, restProps.errors) ?
```
Comment (frassinier) [inline packages/forms/src/UIForm/fieldsets/Tabs.js]: Use classNames here

File: packages/forms/src/UIForm/reducers/form.reducer.js [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+ *          properties: {},
+ *          errors: {},
+ *      },
+ *      ...
+ * }
+ */
+export default function formReducer(state = {}, action) {
+	switch (action.type) {
+	case TF_CREATE_FORM: {
+		const form = state[action.formName];
```
Comment (frassinier) [inline packages/forms/src/UIForm/reducers/form.reducer.js]: Move it up to line 26

File: packages/forms/src/UIForm/utils/properties.js [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+	}
+
+	return key.reduce(
+		(accu, nextKey) => accu && accu[nextKey],
+		properties
+	);
+}
+
+/**
+ * Omit a properties from an object
```
Comment (frassinier) [inline packages/forms/src/UIForm/utils/properties.js]: Omit properties*

**Comment by frassinier** [comment]
```json
{
      "firstname": {
        "type": "string"
      }
}
```
and

 ```json
{
"multipleChoicesList": {
        "type": "array",
        "title": "A multiple choices list",
        "items": {
          "type": "string",
          "enum": [
            "foo",
            "bar",
            "fuzz",
            "qux"
          ]
        },
        "uniqueItems": true
      }
}
```
They could not use sames widgets we need to validate, compute their values and provide fallback widgets before instanciate _/packages/forms/src/UIForm/Widget/Widget.component.js _

File: packages/forms/src/UIForm/Message/Message.component.js [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+		errorMessage,
+		description,
+		isValid,
+	} = props;
+
+	const message = isValid ? description : errorMessage;
+	return message ?
+		(
+			<div
+				className="help-block"
```
Comment (jsomsanith) [inline packages/forms/src/UIForm/Message/Message.component.js]: In fact I'm using [Bootstrap Form Control Validation classes](http://getbootstrap.com/css/#forms-control-validation)

File: packages/forms/src/UIForm/UIForm.component.js [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+		 * User callback: Trigger > after callback.
+		 * Prototype: function onTrigger(type, schema, value, properties)
+		 * This is executed on changes on fields with uiSchema > triggers : ['after']
+		 */
+		onTrigger: PropTypes.func,
+		/** Custom widgets */
+		widgets: PropTypes.object, // eslint-disable-line react/forbid-prop-types
+
+		/** State management impl: The change callback */
+		onChange: PropTypes.func.isRequired,
```
Comment (jsomsanith) [inline packages/forms/src/UIForm/UIForm.component.js]: Because the inputs are controlled inputs. We inject the value, we expect the container to manage the onChange and inject back the new values

File: packages/forms/src/UIForm/actions/constants.js [hunk 16]
```diff
@@ -0,0 +1,6 @@
+export const TF_UPDATE_FORM_DATA = 'TF_UPDATE_FORM_DATA';
+export const TF_SET_ALL_ERRORS = 'TF_SET_ALL_ERRORS';
+export const TF_SET_PARTIAL_ERROR = 'TF_SET_PARTIAL_ERROR';
+export const TF_CREATE_FORM = 'TF_CREATE_FORM';
+export const TF_REMOVE_FORM = 'TF_REMOVE_FORM';
+export const TF_UPDATE_FORM = 'TF_UPDATE_FORM';
```
Comment (jsomsanith) [inline packages/forms/src/UIForm/actions/constants.js]: I put them in an utils/triggers.js and exported/exposed them.

File: packages/forms/src/UIForm/actions/form.actions.test.js (hunk 8 above)
Comment (jsomsanith) [inline packages/forms/src/UIForm/actions/form.actions.test.js]: Snapshots for plain objects are easier to write

File: packages/forms/src/UIForm/__snapshots__/UIForm.test.js.snap [hunk 17]
```diff
@@ -0,0 +1,104 @@
+exports[`UIForm should render 1`] = `
```
Comment (jsomsanith) [inline packages/forms/src/UIForm/__snapshots__/UIForm.test.js.snap]: Removed

File: packages/forms/src/UIForm/UIForm.component.js [hunk 18]
```diff
@@ -0,0 +1,182 @@
...
+	}
+
+	render() {
+		const { autoComplete, errors, formName, id, properties, widgets } = this.props;
+		return (
+			<form
+				onSubmit={this.submit}
+				autoComplete={autoComplete}
+				noValidate
+			>
```
Comment (jsomsanith) [inline packages/forms/src/UIForm/UIForm.component.js]: Let's do that in another PR.
But it will be more easy to extract some rest props and pass them all to the form element

File: packages/forms/src/UIForm/UIForm.component.js [hunk 19]
```diff
@@ -0,0 +1,182 @@
...
+							onChange={this.onChange}
+							onTrigger={this.onTrigger}
+							schema={nextSchema}
+							properties={properties}
+							errors={errors}
+							widgets={widgets}
+						/>
+					))
+				}
+				<button type="submit" className="btn btn-primary">Submit</button>
```
Comment (jsomsanith) [inline packages/forms/src/UIForm/UIForm.component.js]: Let's do that in another PR

File: packages/forms/src/UIForm/UIForm.component.js [hunk 20]
```diff
@@ -0,0 +1,182 @@
...
+	render() {
+		const { autoComplete, errors, formName, id, properties, widgets } = this.props;
+		return (
+			<form
+				onSubmit={this.submit}
+				autoComplete={autoComplete}
+				noValidate
+			>
+				{
+					this.state.mergedSchema.map((nextSchema, index) => (
```
Comment (jsomsanith) [inline packages/forms/src/UIForm/UIForm.component.js]: No it cannot. If the jsonSchema or uiSchema is not correct, the jsfc.merge() fails, this implies a crash.
At this point mergedSchema is always an array

File: packages/forms/src/UIForm/UIForm.component.js [hunk 21]
```diff
@@ -0,0 +1,182 @@
+import React, { PropTypes } from 'react';
+import { merge } from 'talend-json-schema-form-core';
+
+import { validateValue, validateAll } from './utils/validation';
+import Widget from './Widget';
+
+const TRIGGER_AFTER = 'after';
```
Comment (jsomsanith) [inline packages/forms/src/UIForm/UIForm.component.js]: Done
In a trigger.js file, and exported in the index too

File: packages/forms/src/UIForm/UIForm.component.js [hunk 22]
```diff
@@ -0,0 +1,182 @@
...
+		if (!jsonSchema || !uiSchema) {
+			return;
+		}
+		this.setState({
+			mergedSchema: merge(jsonSchema, uiSchema),
+		});
+	}
+
+	/**
+	 * Triggers the onTrigger and onChange if needed
```
Comment (jsomsanith) [inline packages/forms/src/UIForm/UIForm.component.js]: Done

File: packages/forms/src/UIForm/UIForm.component.js [hunk 23]
```diff
@@ -0,0 +1,182 @@
...
+			formName,
+			onChange,
+			properties,
+			customValidation,
+		} = this.props;
+		const error = validateValue(schema, value, properties, customValidation);
+		onChange(formName, schema, value, error);
+
+		const { triggers } = schema;
+		if (triggers && triggers.indexOf(TRIGGER_AFTER) !== -1) {
```
Comment (jsomsanith) [inline packages/forms/src/UIForm/UIForm.component.js]: Done

File: packages/forms/src/UIForm/UIForm.component.js [hunk 24]
```diff
@@ -0,0 +1,182 @@
...
+if (process.env.NODE_ENV !== 'production') {
+	UIForm.propTypes = {
+		/** Form auto complete */
+		autoComplete: PropTypes.bool,
+		/** Form definition: The form name that will be used to create ids */
+		formName: PropTypes.string,
+		/** The form id */
+		id: PropTypes.string,
+		/** Form definition: Json schema that specify the data model */
+		jsonSchema: PropTypes.object, // eslint-disable-line react/forbid-prop-types
```
Comment (jsomsanith) [inline packages/forms/src/UIForm/UIForm.component.js]: Done, set on all the form configurations

File: packages/forms/src/UIForm/Message/Message.component.js [hunk 25]
```diff
@@ -0,0 +1,29 @@
...
+	const {
+		errorMessage,
+		description,
+		isValid,
+	} = props;
+
+	const message = isValid ? description : errorMessage;
+	return message ?
+		(
+			<div
```
Comment (jsomsanith) [inline packages/forms/src/UIForm/Message/Message.component.js]: Done

File: packages/forms/src/UIForm/UIForm.connect.js [hunk 26]
```diff
@@ -0,0 +1,202 @@
...
+import {
+	createForm,
+	removeForm,
+	updateForm,
+	updateFormData,
+	setError,
+	setErrors,
+} from './actions';
+
+class UIForm extends React.Component {
```
Comment (jsomsanith) [inline packages/forms/src/UIForm/UIForm.connect.js]: Done

File: packages/forms/src/UIForm/Widget/Widget.component.js [hunk 27]
```diff
@@ -0,0 +1,50 @@
...
+import { getValue } from '../utils/properties';
+
+export default function Widget(props) {
+	const { errors, formName, onChange, onTrigger, properties, schema, widgets } = props;
+	const { key, type, validationMessage } = schema;
+	const id = sfPath.name(key, '-', formName);
+	const error = errors[key];
+	const errorMessage = validationMessage || error;
+	const WidgetImpl = widgets[type] || defaultWidgets[type];
+	return WidgetImpl ?
```
Comment (jsomsanith) [inline packages/forms/src/UIForm/Widget/Widget.component.js]: Done

File: packages/forms/src/UIForm/fields/Button.js (hunk 9 above)
Comment (jsomsanith) [inline packages/forms/src/UIForm/fields/Button.js]: Done

File: packages/forms/src/UIForm/fields/Button.js (hunk 10 above)
Comment (jsomsanith) [inline packages/forms/src/UIForm/fields/Button.js]: Passing the classname in uiSchema ?
Let's see if we need it first

File: packages/forms/src/UIForm/fieldsets/Tabs.js (hunk 11 above)
Comment (jsomsanith) [inline packages/forms/src/UIForm/fieldsets/Tabs.js]: Done

File: packages/forms/src/UIForm/reducers/form.reducer.js (hunk 12 above)
Comment (jsomsanith) [inline packages/forms/src/UIForm/reducers/form.reducer.js]: Done

File: packages/forms/src/UIForm/utils/properties.js (hunk 13 above)
Comment (jsomsanith) [inline packages/forms/src/UIForm/utils/properties.js]: Done

File: packages/forms/src/UIForm/fields/Button.js (hunk 10 above)
Comment (frassinier) [inline packages/forms/src/UIForm/fields/Button.js]: I thought more about Actions array in fact (like the current impl)
